Fix volume calculation with multiple viewports in 3D

Rookfighter opened the pull request:
* Fixes https://github.com/godotengine/godot/issues/94403
* Based on https://github.com/godotengine/godot/pull/94613

This PR is an up-to-date version of  https://github.com/godotengine/godot/pull/94613 with resolved conflicts.
The `AudioStreamPlayer3D` now searches for the camera / audio listener with the greatest attenuation (typically the closest one) and outputs only the audio associated with that camera. As a result, the audio player always picks the camera where it has the greatest audio output.

Rookfighter committed:
Use camera with highest volume to handle multiple viewports

Co-authored-by: Kusok <118438257+kus04e4ek@users.noreply.github.com>

goatchurchprime commented:
So I thought you could put `output_volume_vector` on the outside of the loop and keep the maximum values in each of its entries as you loop over the cameras.  You're still keeping one direction of the camera from the loop, so this will still have that flipping from left to right in stereo.

For doppler, you could just keep the fastest or closest moving camera.  It's not an effect that is used very much, in my opinion.  I would guess people would want as much doppler as possible, if it is turned on.  

The reason I don't think doppler is much used is that we don't have the capability of setting a low speed of sound so as to exaggerate its effect.  I wanted to do this in a Augmented Reality application, where you are following a spark moving through the room, and the pitch would tell you if it was coming or going from you.  You have to run pretty quickly to notice anything now with the true speed of sound, and it would be very brief.  If I set the speed of sound down to 10m/s then it could come out pretty nice.

In any case, doppler is improperly implemented.  To do it properly there should be a sound buffer for each source and listener to add a delay based on the distance between them.  Then doppler effect is then simply the result of adjusting the length of this buffer as you move closer and further.  And so with two listeners you would get a double sound like an echo where these delay buffers didn't match.  People would probably almost always interpret it as an echo rather than ears being in different places, which means we probably shouldn't do it that way.

Rookfighter commented:
@goatchurchprime  Thanks for the input :+1:
I changed the code now, such that it will use the maximum volume for each channel (left and right) and frame across all cameras. This is pretty similar to the fix which was done in the the [2D audio player](https://github.com/godotengine/godot/pull/76713). I created two temporary vectors, one for the output and one for the reverb volumes. These vectors are used for the individual per camera calculations. I left these calculations unchanged, so they are the same as before. The individual volumes are then fused together by using the maximum from each channel of each frame (for reverb and regular output).
For the Doppler effect I simply use the maximum pitch scale.
I moved all the code which configures the playback and audio buses outside of the camera loop.

AThousandShips commented on scene/3d/audio_stream_player_3d.cpp:
```diff
@@ -354,10 +354,21 @@ StringName AudioStreamPlayer3D::_get_actual_bus() {
 
+static void apply_max_volume_from_vector(Vector<AudioFrame> &tgt_volume_vector, const Vector<AudioFrame> &src_volume_vector) {
+	for (int64_t i = 0; i < tgt_volume_vector.size(); i++) {
+		const auto frame = AudioFrame(
```
```suggestion
		const AudioFrame frame = AudioFrame(
```
`auto` is [not allowed](https://contributing.godotengine.org/en/latest/engine/guidelines/cpp_usage_guidelines.html#auto-keyword)

AThousandShips commented on scene/3d/audio_stream_player_3d.cpp:
```diff
@@ -354,10 +354,21 @@ StringName AudioStreamPlayer3D::_get_actual_bus() {
 	return internal->bus;
 }
 
+static void apply_max_volume_from_vector(Vector<AudioFrame> &tgt_volume_vector, const Vector<AudioFrame> &src_volume_vector) {
```
```suggestion
static void _apply_max_volume_from_vector(Vector<AudioFrame> &r_tgt_volume_vector, const Vector<AudioFrame> &p_src_volume_vector) {
```

AThousandShips replied:
```suggestion
static void _apply_max_volume_from_vector(Vector<AudioFrame> &r_tgt_volume_vector, const Vector<AudioFrame> &p_src_volume_vector) {
```

BuzzLord commented on scene/3d/audio_stream_player_3d.cpp:
```diff
@@ -354,10 +354,21 @@ StringName AudioStreamPlayer3D::_get_actual_bus() {
+	static constexpr int64_t volume_vector_size = 4;
 	Vector<AudioFrame> output_volume_vector;
-	output_volume_vector.resize(4);
+	output_volume_vector.resize(volume_vector_size);
```
The `4` constant comes from the `AudioServer` `MAX_CHANNELS_PER_BUS`, and could be replaced with this (then remove the `volume_vector_size`).

```suggestion
	output_volume_vector.resize(AudioServer::MAX_CHANNELS_PER_BUS);
```

BuzzLord commented on scene/3d/audio_stream_player_3d.cpp:
```diff
@@ -354,10 +354,21 @@ StringName AudioStreamPlayer3D::_get_actual_bus() {
+
 // Interacts with PhysicsServer3D, so can only be called during _physics_process.
 Vector<AudioFrame> AudioStreamPlayer3D::_update_panning() {
+	static constexpr int64_t volume_vector_size = AudioServer::MAX_CHANNELS_PER_BUS;
```
I would change all the `int64_t` variables you've introduced here to just `int`. Anytime I see an int64 I get curious: Why 64? Do you need that precision? Is there something I'm missing? Any edge cases where this could blow up?

In the case of channel counts, not only does it not need 64 bit precision, it could be stored in a char, since I've only ever heard of up to 128 channels in the largest binaural/spatial audio frameworks. Just an int seems fine, for clarity.

```suggestion
	static constexpr int volume_vector_size = AudioServer::MAX_CHANNELS_PER_BUS;
```

Rookfighter replied:
Just to add on this: I was also thinking about this. However, the size and the loop variables are used to resize and index into the `Vector` class. This `Vector` class actually uses `int64_t` as integer type for these operations, so I simply decided to use the target type to avoid unnecessary implicit type conversion all over the place. In this case, any performance gain or memory savings by using a 32-bit type would probably be minor and highly depend on the compiler and the optimization it applies.
So, actually I put some thought into the type I'm using (which is general a good practice) and I would prefer to leave it as `int64_t`

Rookfighter commented on scene/3d/audio_stream_player_3d.cpp:
```diff
@@ -457,28 +478,63 @@ Vector<AudioFrame> AudioStreamPlayer3D::_update_panning() {
 			float tightness = cached_global_panning_strength * 2.0f;
 			tightness *= panning_strength;
-			_calc_output_vol(local_pos.normalized(), tightness, output_volume_vector);
+			_calc_output_vol(local_pos / dist, tightness, tmp_volume_vector);
```
Thanks, you are right! I totally overlooked that the `normalize` method actually performs a check to avoid division by zero :+1: Fixed!

BuzzLord commented on scene/3d/audio_stream_player_3d.cpp:
```diff
@@ -383,8 +394,27 @@ Vector<AudioFrame> AudioStreamPlayer3D::_update_panning() {
+	output_reverb_vector.resize(volume_vector_size);
+	for (AudioFrame &frame : output_volume_vector) {
+		frame = AudioFrame(0, 0);
+	}
```
```suggestion
	Vector<AudioFrame> output_reverb_vector;
	output_reverb_vector.resize(volume_vector_size);
	for (AudioFrame &frame : output_reverb_vector) {
		frame = AudioFrame(0, 0);
	}
```

Rookfighter commented:
@BuzzLord I have addressed your comments, could you have another look if the PR is alright now?

BuzzLord reviewed (changes requested):
I noticed a copy paste bug, but once that's fixed, it looks fine to me. I tested it, and it does what it says.

AThousandShips commented on scene/3d/audio_stream_player_3d.cpp:
```diff
@@ -457,28 +478,63 @@ Vector<AudioFrame> AudioStreamPlayer3D::_update_panning() {
+
+			const Vector3 local_velocity = listener_node->get_global_transform().orthonormalized().basis.xform_inv(linear_velocity - listener_velocity);
+			if (local_velocity != Vector3()) {
+				const float approaching = (local_pos.normalized()).dot(local_velocity.normalized());
```
```suggestion
				const float approaching = local_pos.normalized().dot(local_velocity.normalized());
```

Rookfighter commented on scene/3d/audio_stream_player_3d.cpp:
```diff
@@ -457,28 +478,63 @@ Vector<AudioFrame> AudioStreamPlayer3D::_update_panning() {
+
+			const Vector3 local_velocity = listener_node->get_global_transform().orthonormalized().basis.xform_inv(linear_velocity - listener_velocity);
+			if (local_velocity != Vector3()) {
+				const float approaching = (local_pos.normalized()).dot(local_velocity.normalized());
```
Thanks, changed :+1:

BuzzLord reviewed (changes requested):
While testing with doppler, I noticed it tended to keep higher pitch more than lower pitch. While looking to see if there was an better way to blend pitches, I realized there's a bug: `actual_pitch_scale` can never be less than `internal->pitch_scale` because of the `MAX`. So you never get low pitches for receding sounds, even in the single camera case.

I think I came up with a better way of calculating the actual pitch, using a weighted average. Since humans here pitch roughly in log2 space, the idea is to keep weighted log2 pitch scales, then normalize and change back to linear at the end.

```
	// set picth to default value
	// this will only change if the Doppler effect is activated
	actual_pitch_scale = internal->pitch_scale;
	float log_pitch_scale = 0.0;
	float log_pitch_weight = 0.0;
```

Then in the doppler section, instead of `MAX` you do:
```
	float doppler_pitch_scale = internal->pitch_scale * speed_of_sound / (speed_of_sound + velocity * approaching);
	doppler_pitch_scale = CLAMP(doppler_pitch_scale, (1 / 8.0), 8.0); //avoid crazy stuff

	float weight = _get_max_volume(tmp_volume_vector);
	log_pitch_scale += weight * log2(doppler_pitch_scale);
	log_pitch_weight += weight;
```

Where `_get_max_volume` just gets the largest volume response of any channel:
```
static float _get_max_volume(const Vector<AudioFrame> &p_src_volume_vector) {
	float max_vol = 0.0;
	for (int64_t i = 0; i < p_src_volume_vector.size(); i++) {
		max_vol = MAX(max_vol, p_src_volume_vector[i].left);
		max_vol = MAX(max_vol, p_src_volume_vector[i].right);
	}
	return max_vol;
}
```

After the cameras loop:
```
	if (log_pitch_weight > 0) {
		actual_pitch_scale = pow(2.0, log_pitch_scale / log_pitch_weight);
	}
```

BuzzLord commented:
> Your approach with a weighted average sounds alright to me, not sure why the logarithmic scale makes sense but I guess its kind of like logarithmic gamma correction in images (I'm actually more an image than an audio person 😄 ).

The way I was thinking about it was, if you have one sound that is one octave up (i.e. 2.0x samples per second as normal) and another sound at one octave down (i.e. 0.5x samples per second), an even blend of the two should be just the 1x normal sample rate.

Those multipliers are what the pitch scale is. If you averaged them in linear space, you get `(2.0 + 0.5) / 2 = 1.25`. If you average them in log space you get `(log2(2.0) + log2(0.5)) / 2 = (1.0 + (-1.0)) / 2 = 0.0`, then when you bring them back to linear value, `pow(2, 0.0) = 1.0`.

Rookfighter commented:
@BuzzLord Thanks for the explanation and sorry for the long response time (I have been on a business trip). Makes totally sense to me :+1: I implemented the pitch scale calculation as suggested. It seems to do what it should, but do you have a dedicated project / repo where you tested the Doppler effect?

BuzzLord reviewed (commented):
Something that came up in the last audio meeting (that I forgot to mention before) is it's not ideal to allocate `tmp_volume_vector` and `tmp_reverb_vector` every frame. The `output_volume_vector` makes sense, since it gets returned, but these are only used in this function, and not passed anywhere. 

I think since they're needed every frame, you could move them to class member variables, and just have them resize to the right size at the start of `_update_panning`. Or allocating them on the stack with `AudioFrame tmp_volume_vector[4];` would probably be fine too, but would require some refactoring to make the functions that use them work. I'm not sure which is best; maybe we can discuss it at the next meeting.

I'd also rename them to something like `listener_volume_vector`/`listener_reverb_vector`, or `camera_` or `view_`, or something to that effect.

Rookfighter commented:
@BuzzLord Thanks for the input and I agree we should avoid the overhead of allocating and freeing the heap memory every iteration. I played around with [`FixedVector`(https://github.com/godotengine/godot/blob/master/core/templates/fixed_vector.h) instead of plain C-style arrays, but I would still have to adjust various function signatures. In the end I decided to move the tmp vectors to class members as it seems cleaner in terms of changes and existing interface, but really it is a matter taste IMO.

BuzzLord reviewed (approved):
Looks good to me. Tested it, and it works as intended. I think the implementation is straight forward and clean, and doesn't affect anything outside of the use-case it fixes (i.e. getting better 3d panning/mixing with multiple cameras/listeners).

I'm honestly not a fan of the chunky three line, triple /// comment style used in the header (just aesthetically), but there is precedence elsewhere in the engine for that type of comment, and the comment content is fine, so I guess it's fine.

Ivorforce commented on scene/3d/audio_stream_player_3d.cpp:
```diff
@@ -354,10 +354,31 @@ StringName AudioStreamPlayer3D::_get_actual_bus() {
+				MAX(r_tgt_volume_vector[i].right, p_src_volume_vector[i].right));
+
+		r_tgt_volume_vector.write[i] = frame;
+	}
```
An indexed loop is slow with `Vector` and thus discouraged. Use pointers instead.

```suggestion
	AudioFrame *tgt_ptr = r_tgt_volume_vector.ptrw();
	const AudioFrame *src_ptr = p_src_volume_vector.ptr();
	
	for (int64_t i = 0; i < r_tgt_volume_vector.size(); i++) {
		const AudioFrame frame = AudioFrame(
				MAX(src_ptr[i].left, src_ptr[i].left),
				MAX(tgt_ptr[i].right, src_ptr[i].right));

		tgt_ptr[i] = frame;
	}
```

Ivorforce reviewed (commented):
I need to do another review pass, but here's some quick feedback.

Rookfighter commented on scene/3d/audio_stream_player_3d.cpp:
```diff
@@ -354,10 +354,31 @@ StringName AudioStreamPlayer3D::_get_actual_bus() {
+		max_vol = MAX(max_vol, p_src_volume_vector[i].right);
+	}
+	return max_vol;
+}
```
Thanks, actually I should use the iterator whenever possible in the first place.

Rookfighter commented on scene/3d/audio_stream_player_3d.cpp:
```diff
@@ -457,28 +486,71 @@ Vector<AudioFrame> AudioStreamPlayer3D::_update_panning() {
+		AudioServer::get_singleton()->set_playback_highshelf_params(playback, linear_attenuation, attenuation_filter_cutoff_hz);
+	}
+
+	HashMap<StringName, Vector<AudioFrame>> bus_volumes;
```
Good point, actually the temporary hash map has been there before, so there is no performance loss introduced here (compared to before). The `HashMap` type itself is defined by `set_playback_bus_volumes_linear` of the `AudioServer` class, so I cannot switch to a stack only alternative. However, I figured that we can make the hash map a class member which should now avoid unnecessary heap allocations.

BuzzLord replied:
It is called each physics frame (that the sound is active). There used to be a way for it to only update when needed, but there is a bug that causes that to not happen.

The bug is caused by `last_mix_count` never being updated; the line that updated it got lost in a refactor some time ago.

Rookfighter commented:
@Ivorforce Thanks for your review. I either fixed or commented on all your findings :+1:
For the local vectors I changed from `Vector` to `FixedVector` and allocate them now locally on the stack (their capacity is only `4`), instead of switching to `LocalVector`. If you do not like this change, just let me know and I will revert it and use `LocalVector` class members for the temporaries.

Ivorforce commented on scene/3d/audio_stream_player_3d.h:
```diff
@@ -66,6 +67,8 @@ class AudioStreamPlayer3D : public Node3D {
 
 	};
 
+	static constexpr int64_t volume_vector_size = AudioServer::MAX_CHANNELS_PER_BUS;
```
We use `UPPER_SNAKE_CASE` for constants

```suggestion
	static constexpr int64_t VOLUME_VECTOR_SIZE = AudioServer::MAX_CHANNELS_PER_BUS;
```

Ivorforce commented on scene/3d/audio_stream_player_3d.cpp:
```diff
@@ -488,46 +559,24 @@ Vector<AudioFrame> AudioStreamPlayer3D::_update_panning() {
 		}
+	}
+
+	// if no listeners are in range and this was the case last frame, then we can skip setting any audio
```
Use sentence case for comments, ending with a dot:

```suggestion
	// If no listeners are in range and this was the case last frame, then we can skip setting any audio.
```

Ivorforce reviewed (commented):
Mostly stylistic comments now, most of your changes look great!

BuzzLord commented on scene/3d/audio_stream_player_3d.cpp:
```diff
@@ -457,28 +486,71 @@ Vector<AudioFrame> AudioStreamPlayer3D::_update_panning() {
+				static constexpr float speed_of_sound = 343.0F;
+
+				float doppler_pitch_scale = internal->pitch_scale * speed_of_sound / (speed_of_sound + velocity * approaching);
+				doppler_pitch_scale = CLAMP(doppler_pitch_scale, (1.0F / 8.0F), 8.0F); //avoid crazy stuff
```
When testing this, I found that audio started to sound pretty bad (due to very high/low freqs, and distortion caused by resampling) long before it reached these limits, so I think they're reasonable.

Ivorforce reviewed (approved):
LGTM.

Fairly complex change, but i agree with the problem, the fix approach, and the implementation.

I assume if any problems arise from the change we will still have time to fix them before final release.

Thank you for all the hard work!! (and thanks to @BuzzLord too for the review and testing!)
